pkg/controller/operands/ssp.go

```diff
@@ -3,289 +3,125 @@ package operands
-
-func (h nodeLabellerBundleHooks) getFullCr(hc *hcov1beta1.HyperConverged) runtime.Object {
-	return NewKubeVirtNodeLabellerBundleForCR(hc, hc.Namespace)
-}
-func (h nodeLabellerBundleHooks) getEmptyCr() runtime.Object {
-	return &sspv1.KubevirtNodeLabellerBundle{}
-}
-func (h nodeLabellerBundleHooks) validate() error                                        { return nil }
-func (h nodeLabellerBundleHooks) postFound(_ *common.HcoRequest, _ runtime.Object) error { return nil }
-func (h nodeLabellerBundleHooks) getConditions(cr runtime.Object) []conditionsv1.Condition {
-	return cr.(*sspv1.KubevirtNodeLabellerBundle).Status.Conditions
-}
-func (h nodeLabellerBundleHooks) checkComponentVersion(cr runtime.Object) bool {
-	found := cr.(*sspv1.KubevirtNodeLabellerBundle)
-	return checkComponentVersion(hcoutil.SspVersionEnvV, found.Status.ObservedVersion)
-}
-func (h nodeLabellerBundleHooks) getObjectMeta(cr runtime.Object) *metav1.ObjectMeta {
-	return &cr.(*sspv1.KubevirtNodeLabellerBundle).ObjectMeta
-}
-
-func (h *nodeLabellerBundleHooks) updateCr(req *common.HcoRequest, Client client.Client, exists runtime.Object, required runtime.Object) (bool, bool, error) {
-	kvNLB, ok1 := required.(*sspv1.KubevirtNodeLabellerBundle)
-	found, ok2 := exists.(*sspv1.KubevirtNodeLabellerBundle)
-	if !ok1 || !ok2 {
-		return false, false, errors.New("can't convert to KubeVirt Node Labeller Bundle")
-	}
-	if !reflect.DeepEqual(kvNLB.Spec, found.Spec) {
-		if req.HCOTriggered {
-			req.Logger.Info("Updating existing KubeVirt Node Labeller Bundle's Spec to new opinionated values")
-		} else {
-			req.Logger.Info("Reconciling an externally updated KubeVirt Node Labeller Bundle's Spec to its opinionated values")
-		}
-		kvNLB.Spec.DeepCopyInto(&found.Spec)
-		err := Client.Update(req.Ctx, found)
-		if err != nil {
-			return false, false, err
-		}
 
-		return true, !req.HCOTriggered, nil
+		crdsToRemove: append([]string{}, oldSSPCrds...),
 	}
-	return false, false, nil
 }
 
-func (h nodeLabellerBundleHandler) Ensure(req *common.HcoRequest) *EnsureResult {
-	handler := sspOperand(h)
-	return handler.ensure(req)
-}
-
-func NewKubeVirtNodeLabellerBundleForCR(cr *hcov1beta1.HyperConverged, namespace string) *sspv1.KubevirtNodeLabellerBundle {
-	labels := map[string]string{
-		hcoutil.AppLabel: cr.Name,
-	}
+func (handler *sspHandler) ensure(req *common.HcoRequest) *EnsureResult {
+	res := handler.genericOperand.ensure(req)
 
-	spec := sspv1.ComponentSpec{
-		// UseKVM: isKVMAvailable(),
-	}
+	// Attempt to remove old CRDs
+	if len(handler.crdsToRemove) > 0 && (!req.UpgradeMode || res.Updated) {
+		unremovedCRDs := make([]string, 0, len(handler.crdsToRemove))
 
-	if cr.Spec.Workloads.NodePlacement != nil {
-		if cr.Spec.Workloads.NodePlacement.Affinity != nil {
-			cr.Spec.Workloads.NodePlacement.Affinity.DeepCopyInto(&spec.Affinity)
-		}
+		for _, crdName := range handler.crdsToRemove {
+			removed := removeCrd(handler.Client, req, crdName)
```

Comment:
Yes this cries out for concurrency. I think this is simple enough so better be implemented now rather than postponed to sometimes between [unknown, never].

Reply:
Done.

Reply:
By the way, the way I implemented this assumes that the controller-runtime's Client is thread-safe. I'm not sure whether this assumption is correct or not, but [asked](https://groups.google.com/u/3/g/kubebuilder/c/CT0wdZykPwc).

Reply:
Here's the answer I got from one of the controller-runtime devs (for some reason it doesn't appear on the google group thread):

_Yes, the delegating client is safe to use concurrently. The delegating client itself doesn't have any moving parts after its creation, and the underlying clients (cached, and live) should be already safe._